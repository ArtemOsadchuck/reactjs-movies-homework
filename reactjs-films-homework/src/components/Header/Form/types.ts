export interface IFormProps {
  placeholder: string;
}
