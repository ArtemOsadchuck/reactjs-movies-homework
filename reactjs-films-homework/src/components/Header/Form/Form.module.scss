.form {
  align-items: center;
  display: flex;
  justify-content: flex-end;
  // padding-right: 15%;
  margin-right: 20%;
  position: relative;
  text-align: start;

  @media screen and (max-width: 760px) {
    margin-right: 0.5rem;
  }

  .input {
    border: 0;
    border-radius: 3px;
    height: 2rem;
    padding-left: 0.5rem;
    text-align: start;
    width: 28rem;

    @media screen and (max-width: 760px) {
      width: 18rem;
    }

    @media screen and (max-width: 425px) {
      width: 11rem;
    }
  }

  .search {
    align-items: center;
    background-color: #fff;
    display: flex;
    height: 80%;
    justify-content: center;
    margin: auto;
    position: absolute;
    right: 0.1rem;
    width: 2rem;

    &:hover {
      cursor: pointer;
    }
  }
}

.formWrapper {
  @media screen and (max-width: 760px) {
    flex-grow: 3;
  }
}
