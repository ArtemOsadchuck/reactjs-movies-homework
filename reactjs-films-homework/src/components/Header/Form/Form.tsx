import React, { useEffect, useState } from 'react';
import search from './assets/search.png';
import styles from './Form.module.scss';
import { useAppDispatch } from '../../../hooks/hooks';
import {
  setQuery,
  setActivePage,
} from '../../../store/rootStore/mainStore/mainSlice';

import { Link, useHistory } from 'react-router-dom';
import useUrlSearch from '../../../hooks/useUrlSearch';

type IFormProps = {
  placeholder: string;
};

const Form: React.FC<IFormProps> = ({ placeholder }) => {
  const [inputValue, setInputValue] = useState<string>('');
  const [inputSearchValue, setInputSearchValue] = useState<string>('');

  const pageAfterSearch = '1';

  const dispatch = useAppDispatch();
  const history = useHistory();
  const activeUrlCategory = useUrlSearch('category');
  const activeUrlPage = useUrlSearch('page');
  const activeUrlSearch = useUrlSearch('search');

  const [placeholderState, setPlaceholderState] = useState<string>(placeholder);

  const resetInput = () => {
    const checkIsOnBlur = document.body.addEventListener(
      'click',
      (event: any) => {
        const id: string = event.target.id;
        if (id === 'search') {
          setInputValue('');
        }
        if (id === 'input_search') {
          return;
        }
        setInputValue('');
      },
      { once: true }
    );
    return checkIsOnBlur;
  };

  useEffect(() => {
    setPlaceholderState((prev) => {
      return (prev = !activeUrlSearch
        ? placeholder
        : `Results for ${activeUrlSearch}`);
    });

    activeUrlPage && dispatch(setActivePage(pageAfterSearch));
    activeUrlSearch && dispatch(setQuery(activeUrlSearch));
  }, [activeUrlPage, activeUrlSearch, dispatch, placeholder]);

  useEffect(() => {
    activeUrlCategory && setPlaceholderState((prev) => placeholder);
    setInputSearchValue('');
  }, [placeholder, activeUrlCategory]);

  const getSearchResults = () => {
    setPlaceholderState((prev) => {
      return (prev = !inputValue ? placeholder : `Results for ${inputValue}`);
    });

    dispatch(setActivePage(pageAfterSearch));
    dispatch(setQuery(inputValue));
    setInputValue('');
    setInputSearchValue('');
  };

  return (
    <div className={styles.formWrapper} onBlur={resetInput}>
      <form className={styles.form} id={'form'}>
        <input
          type="search"
          className={styles.input}
          placeholder={placeholderState}
          id="input_search"
          autoComplete="off"
          onChange={(event) => {
            setInputValue((prev) => (prev = event.target.value));
            setInputSearchValue((prev) => (prev = event.target.value));
          }}
          onKeyPress={(event) => {
            if (event.code === 'Enter') {
              event.preventDefault();
              getSearchResults();
              history.push(
                `/?search=${inputSearchValue}&page=${pageAfterSearch}`
              );
            }
          }}
          value={inputValue}
        />
        <div
          className={styles.search}
          id={'search'}
          onClick={(event) => {
            event.preventDefault();
            getSearchResults();
          }}
        >
          <Link to={`/?search=${inputSearchValue}&page=${pageAfterSearch}`}>
            <img src={search} id={'search'} alt="" width="16px" />
          </Link>
        </div>
      </form>
    </div>
  );
};

export default Form;
