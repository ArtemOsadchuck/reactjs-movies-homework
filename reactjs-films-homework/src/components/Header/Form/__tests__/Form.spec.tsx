import React from 'react';
import { render, cleanup } from '@testing-library/react';

import Form from '../Form';

import RouterWrapper from '../../../../__testsUtils__/routerHoc';
import StoreWrapper from '../../../../__testsUtils__/storeHoc';

const renderComponent = (placeholder: string) => {
  const urlApp = '/';
  const { asFragment } = render(
    <StoreWrapper>
      <RouterWrapper url={urlApp}>
        <Form placeholder="Movies" />
      </RouterWrapper>
    </StoreWrapper>
  );
  return asFragment();
};

describe('Form', () => {
  it('should renders correctly', () => {
    const placeholder = 'Movies';
    const fragment = renderComponent(placeholder);

    expect(fragment).toMatchSnapshot();
  });

  cleanup();
});
