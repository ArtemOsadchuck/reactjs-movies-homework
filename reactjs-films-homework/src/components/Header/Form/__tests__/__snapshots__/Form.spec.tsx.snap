// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form Snapshot 1`] = `
<DocumentFragment>
  <div
    class="formWrapper"
  >
    <form
      class="form"
      id="form"
    >
      <input
        autocomplete="off"
        class="input"
        id="input_search"
        placeholder="Movies"
        type="search"
        value=""
      />
      <div
        class="search"
        id="search"
      >
        <img
          alt=""
          id="search"
          src="search.png"
          width="16px"
        />
      </div>
    </form>
  </div>
</DocumentFragment>
`;
